File: push/src/main/java/com/xiaomi/tinyData/TinyDataCacheReader.java
package com.xiaomi.tinyData;

import android.content.Context;
import android.content.SharedPreferences;
import com.xiaomi.channel.commonutils.android.DataCryptUtils;
import com.xiaomi.channel.commonutils.file.IOUtils;
import com.xiaomi.channel.commonutils.logger.MyLog;
import com.xiaomi.channel.commonutils.misc.ByteUtils;
import com.xiaomi.channel.commonutils.misc.ScheduledJobManager;
import com.xiaomi.push.service.TinyDataStorage;
import com.xiaomi.xmpush.thrift.ClientUploadDataItem;
import com.xiaomi.xmpush.thrift.XmPushThriftSerializeUtils;
import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.channels.FileLock;
import java.util.ArrayList;
import java.util.List;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/tinyData/TinyDataCacheReader.class */
public class TinyDataCacheReader {
    private static final int TINY_DATA_MAX_UPLOAD_ITEM_COUNT = 8;
    private static final String TINY_DATA_READ_TEMP_FILE_DIR = "/tdReadTemp";
    private static boolean mTinyDataJobIsRunning = false;

    /* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/tinyData/TinyDataCacheReader$TinyDataReadJob.class */
    static class TinyDataReadJob implements Runnable {
        private Context mContext;
        private TinyDataUploader mUploader;

        public TinyDataReadJob(Context context, TinyDataUploader tinyDataUploader) {
            this.mUploader = tinyDataUploader;
            this.mContext = context;
        }

        @Override // java.lang.Runnable
        public void run() {
            try {
                TinyDataCacheReader.extractTinyData(this.mContext, this.mUploader);
            } catch (Throwable th) {
                MyLog.e(th);
            }
        }
    }

    public static void addTinyDataCacheReadJob(Context context, TinyDataUploader tinyDataUploader) {
        ScheduledJobManager.getInstance(context).addOneShootJob(new TinyDataReadJob(context, tinyDataUploader));
    }

    /* JADX INFO: Access modifiers changed from: private */
    public static void extractTinyData(Context context, TinyDataUploader tinyDataUploader) throws Throwable {
        if (mTinyDataJobIsRunning) {
            MyLog.w("TinyData extractTinyData is running");
            return;
        }
        mTinyDataJobIsRunning = true;
        File file = new File(context.getFilesDir(), TinyDataStorage.TINY_DATA_CACHE_FILE_NAME);
        if (!file.exists()) {
            MyLog.w("TinyData no ready file to get data.");
            return;
        }
        verifyFileDir(context);
        byte[] tinyDataKeyWithDefault = TinyDataStorage.getTinyDataKeyWithDefault(context);
        RandomAccessFile randomAccessFile = null;
        FileLock fileLock = null;
        try {
            File file2 = new File(context.getFilesDir(), TinyDataStorage.TINY_DATA_CACHE_FILE_LOCK);
            IOUtils.createFileQuietly(file2);
            randomAccessFile = new RandomAccessFile(file2, "rw");
            fileLock = randomAccessFile.getChannel().lock();
            File file3 = new File(context.getFilesDir() + TINY_DATA_READ_TEMP_FILE_DIR + "/" + TinyDataStorage.TINY_DATA_CACHE_FILE_NAME);
            file.renameTo(file3);
        } catch (Exception e) {
            MyLog.e(e);
        } finally {
            if (fileLock != null && fileLock.isValid()) {
                try {
                    fileLock.release();
                } catch (IOException e2) {
                    MyLog.e(e2);
                }
            }
            IOUtils.closeQuietly(randomAccessFile);
        }
        try {
            File file3 = new File(context.getFilesDir() + TINY_DATA_READ_TEMP_FILE_DIR + "/" + TinyDataStorage.TINY_DATA_CACHE_FILE_NAME);
            if (!file3.exists()) {
                MyLog.w("TinyData no ready file to get data.");
                return;
            }
            readTinyDataFromFile(context, tinyDataUploader, file3, tinyDataKeyWithDefault);
            updateTinyDataUploadTimeStamp(context);
        } catch (Throwable th) {
            throw th;
        } finally {
            TinyDataCacheProcessor.setIsTinyDataExtracting(false);
            mTinyDataJobIsRunning = false;
        }
    }

    private static void readTinyDataFromFile(Context context, TinyDataUploader tinyDataUploader, File file, byte[] bArr) throws Throwable {
        BufferedInputStream bufferedInputStream = null;
        List<ClientUploadDataItem> arrayList = new ArrayList<>(TINY_DATA_MAX_UPLOAD_ITEM_COUNT);
        try {
            bufferedInputStream = new BufferedInputStream(new FileInputStream(file));
            byte[] bArr2 = new byte[4];
            while (true) {
                int read = bufferedInputStream.read();
                if (read == -1) {
                    break;
                }
                bArr2[0] = (byte) read;
                if (bufferedInputStream.read(bArr2, 1, 3) != 3) {
                    MyLog.e("TinyData read from cache file failed cause lengthBuffer error.");
                    break;
                }
                int i = ByteUtils.toInt(bArr2);
                if (i < 1 || i > 10240) {
                    MyLog.e("TinyData read from cache file failed cause lengthBuffer < 1 || too big. length:" + i);
                    break;
                }
                byte[] bArr3 = new byte[i];
                int i2 = 0;
                while (i2 < i) {
                    int read2 = bufferedInputStream.read(bArr3, i2, i - i2);
                    if (read2 == -1) {
                        break;
                    }
                    i2 += read2;
                }
                if (i2 != i) {
                    MyLog.e("TinyData read from cache file failed cause buffer size not equal length. size:" + i2 + "__length:" + i);
                    break;
                }
                try {
                    byte[] mipushDecrypt = DataCryptUtils.mipushDecrypt(bArr, bArr3);
                    ClientUploadDataItem clientUploadDataItem = new ClientUploadDataItem();
                    XmPushThriftSerializeUtils.convertByteArrayToThriftObject(clientUploadDataItem, mipushDecrypt);
                    arrayList.add(clientUploadDataItem);
                    if (arrayList.size() >= TINY_DATA_MAX_UPLOAD_ITEM_COUNT) {
                        TinyDataCacheUploader.uploadTinyData(context, tinyDataUploader, arrayList);
                        arrayList = new ArrayList<>(TINY_DATA_MAX_UPLOAD_ITEM_COUNT);
                    }
                } catch (Exception e) {
                    MyLog.e(e);
                }
            }
            if (!arrayList.isEmpty()) {
                TinyDataCacheUploader.uploadTinyData(context, tinyDataUploader, arrayList);
            }
        } finally {
            IOUtils.closeQuietly(bufferedInputStream);
            if (file.exists() && !file.delete()) {
                MyLog.w("TinyData delete reading temp file failed");
            }
        }
    }

    private static void updateTinyDataUploadTimeStamp(Context context) {
        SharedPreferences.Editor editorEdit = context.getSharedPreferences("mipush_extra", 4).edit();
        editorEdit.putLong(TinyDataCacheProcessor.LAST_TINY_DATA_UPLOAD_TIMESTAMP, System.currentTimeMillis() / 1000);
        editorEdit.commit();
    }

    private static void verifyFileDir(Context context) {
        File file = new File(context.getFilesDir() + TINY_DATA_READ_TEMP_FILE_DIR);
        if (file.exists()) {
            return;
        }
        file.mkdirs();
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/tinyData/TinyDataCacheReader.kt
package com.xiaomi.tinyData

import android.content.Context
import com.xiaomi.channel.commonutils.android.DataCryptUtils
import com.xiaomi.channel.commonutils.file.IOUtils
import com.xiaomi.channel.commonutils.logger.LoggerInterface
import com.xiaomi.channel.commonutils.logger.MyLog
import com.xiaomi.channel.commonutils.misc.ByteUtils
import com.xiaomi.channel.commonutils.misc.ScheduledJobManager
import com.xiaomi.push.service.TinyDataStorage
import com.xiaomi.xmpush.thrift.ClientUploadDataItem
import com.xiaomi.xmpush.thrift.XmPushThriftSerializeUtils
import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.io.RandomAccessFile
import java.nio.channels.FileLock

class TinyDataCacheReader private constructor() {
    private class TinyDataReadJob(
        private val mContext: Context,
        private val mUploader: TinyDataUploader?
    ) : Runnable {
        override fun run() {
            try {
                extractTinyData(mContext, mUploader)
            } catch (th: Throwable) {
                MyLog.e(th)
            }
        }
    }

    companion object {
        private const val TINY_DATA_MAX_UPLOAD_ITEM_COUNT = 8
        private const val TINY_DATA_READ_TEMP_FILE_DIR = "/tdReadTemp"
        private var mTinyDataJobIsRunning = false

        fun addTinyDataCacheReadJob(context: Context, tinyDataUploader: TinyDataUploader?) {
            ScheduledJobManager.getInstance(context).addOneShootJob(TinyDataReadJob(context, tinyDataUploader))
        }

        @Throws(Throwable::class)
        private fun extractTinyData(context: Context, tinyDataUploader: TinyDataUploader?) {
            if (mTinyDataJobIsRunning) {
                MyLog.w("TinyData extractTinyData is running")
                return
            }
            mTinyDataJobIsRunning = true
            val file = File(context.filesDir, TinyDataStorage.TINY_DATA_CACHE_FILE_NAME)
            if (!file.exists()) {
                MyLog.w("TinyData no ready file to get data.")
                return
            }
            verifyFileDir(context)
            val tinyDataKeyWithDefault = TinyDataStorage.getTinyDataKeyWithDefault(context)
            var randomAccessFile: RandomAccessFile? = null
            var fileLock: FileLock? = null
            try {
                val lockFile = File(context.filesDir, TinyDataStorage.TINY_DATA_CACHE_FILE_LOCK)
                IOUtils.createFileQuietly(lockFile)
                randomAccessFile = RandomAccessFile(lockFile, "rw")
                fileLock = randomAccessFile.channel.lock()
                val tempFile = File(context.filesDir.toString() + TINY_DATA_READ_TEMP_FILE_DIR + "/" + TinyDataStorage.TINY_DATA_CACHE_FILE_NAME)
                file.renameTo(tempFile)
            } catch (e: Exception) {
                MyLog.e(e)
            } finally {
                if (fileLock != null && fileLock.isValid) {
                    try {
                        fileLock.release()
                    } catch (e: java.io.IOException) {
                        MyLog.e(e)
                    }
                }
                IOUtils.closeQuietly(randomAccessFile)
            }
            try {
                val tempFile = File(context.filesDir.toString() + TINY_DATA_READ_TEMP_FILE_DIR + "/" + TinyDataStorage.TINY_DATA_CACHE_FILE_NAME)
                if (!tempFile.exists()) {
                    MyLog.w("TinyData no ready file to get data.")
                    return
                }
                readTinyDataFromFile(context, tinyDataUploader, tempFile, tinyDataKeyWithDefault)
                updateTinyDataUploadTimeStamp(context)
            } catch (th: Throwable) {
                throw th
            } finally {
                TinyDataCacheProcessor.setIsTinyDataExtracting(false)
                mTinyDataJobIsRunning = false
            }
        }

        @Throws(Throwable::class)
        private fun readTinyDataFromFile(
            context: Context,
            tinyDataUploader: TinyDataUploader?,
            file: File,
            key: ByteArray
        ) {
            var bufferedInputStream: BufferedInputStream? = null
            val arrayList = ArrayList<ClientUploadDataItem>(TINY_DATA_MAX_UPLOAD_ITEM_COUNT)
            try {
                bufferedInputStream = BufferedInputStream(FileInputStream(file))
                val lengthBuffer = ByteArray(4)
                while (true) {
                    val read = bufferedInputStream.read()
                    if (read == -1) break
                    lengthBuffer[0] = read.toByte()
                    if (bufferedInputStream.read(lengthBuffer, 1, 3) != 3) {
                        MyLog.e("TinyData read from cache file failed cause lengthBuffer error.")
                        break
                    }
                    val length = ByteUtils.toInt(lengthBuffer)
                    if (length < 1 || length > 10240) {
                        MyLog.e("TinyData read from cache file failed cause lengthBuffer < 1 || too big. length:$length")
                        break
                    }
                    val buffer = ByteArray(length)
                    var bytesRead = 0
                    while (bytesRead < length) {
                        val read2 = bufferedInputStream.read(buffer, bytesRead, length - bytesRead)
                        if (read2 == -1) break
                        bytesRead += read2
                    }
                    if (bytesRead != length) {
                        MyLog.e("TinyData read from cache file failed cause buffer size not equal length. size:${bytesRead}__length:$length")
                        break
                    }
                    try {
                        val decrypted = DataCryptUtils.mipushDecrypt(key, buffer)
                        val item = ClientUploadDataItem()
                        XmPushThriftSerializeUtils.convertByteArrayToThriftObject(item, decrypted)
                        arrayList.add(item)
                        if (arrayList.size >= TINY_DATA_MAX_UPLOAD_ITEM_COUNT) {
                            TinyDataCacheUploader.uploadTinyData(context, tinyDataUploader, arrayList)
                            arrayList.clear()
                        }
                    } catch (e: Exception) {
                        MyLog.e(e)
                    }
                }
                if (arrayList.isNotEmpty()) {
                    TinyDataCacheUploader.uploadTinyData(context, tinyDataUploader, arrayList)
                }
            } finally {
                IOUtils.closeQuietly(bufferedInputStream)
                if (file.exists() && !file.delete()) {
                    MyLog.w("TinyData delete reading temp file failed")
                }
            }
        }

        private fun updateTinyDataUploadTimeStamp(context: Context) {
            context.getSharedPreferences("mipush_extra", 4).edit()
                .putLong(TinyDataCacheProcessor.LAST_TINY_DATA_UPLOAD_TIMESTAMP, System.currentTimeMillis() / 1000)
                .commit()
        }

        private fun verifyFileDir(context: Context) {
            val dir = File(context.filesDir.toString() + TINY_DATA_READ_TEMP_FILE_DIR)
            if (!dir.exists()) {
                dir.mkdirs()
            }
        }
    }
}
